### PR TITLE
[MRG] ENH: Add Basic Gaussian Process Interface

### PR DESCRIPTION
**Pull request by tirthasheshpatel**
According to the discussion on [this discourse thread](https://discourse.pymc.io/t/adding-gaussian-processes-in-pymc4-this-summer/4538), I have created a simple GP interface for PyMC4.

- [x] add base classes for covariance functions, mean functions, gp.
- [x] add a working ``LatentGP`` model for univariate and multivariate data.
- [x] add tests for ``LatentGP``
- [x] add tests for mean functions and covariance functions
- [x] add tests for a combination of mean and covariance functions
- [x] ~add an interface for combining GP models~
- [x] ~error handling for invalid shapes (maybe?)~
- [x] add a notebook with pretty plots.
- [x] add ``__radd__`` and ``__rmul__`` methods for combination of covariance functions with non-covariance functions.
- [x] add tests for ``__radd__`` and ``__rmul__`` methods.
- [x] add typing in all files

Thanks, @fonnesbeck for your suggestions and help! I think this a better place for discussion for further development.
Comments and criticism are welcome!

### CODE REVIEW

**Comment by twiecki** [comment]
@tirthasheshpatel This looks like a great start -- thanks for taking this on! @fonnesbeck and @bwengals are good people to provide some feedback. I think a short NB that demonstrates the functionality so far would also go a long way.

**Comment by tirthasheshpatel** [comment]
Thanks, @twiecki. I will add a notebook soon!

**Comment by twiecki** [comment]
@tirthasheshpatel Looks great -- bonus points for pretty plots ;). You can probably also just port existing examples from PyMC3.

**Comment by tirthasheshpatel** [comment]
Hey @fonnesbeck @aloctavodia @bwengals,

I think that this pr is almost ready for review. Can you please review this and suggest any further changes? Thanks!

<details>
<summary>I don't know why are the codecov reports aren't uploaded</summary>

```none
----------- coverage: platform win32, python 3.7.3-final-0 -----------
Name                                          Stmts   Miss  Cover
-----------------------------------------------------------------
pymc4\__init__.py                                13      0   100%
pymc4\coroutine_model.py                         59      6    90%
pymc4\distributions\__init__.py                   6      0   100%
pymc4\distributions\batchstack.py                72     20    72%
pymc4\distributions\continuous.py               167      2    99%
pymc4\distributions\discrete.py                  87      2    98%
pymc4\distributions\distribution.py             145     17    88%
pymc4\distributions\half_student_t.py           107     39    64%
pymc4\distributions\mixture.py                    4      4     0%
pymc4\distributions\multivariate.py              49      0   100%
pymc4\distributions\timeseries.py                19      0   100%
pymc4\distributions\transforms.py                58     14    76%
pymc4\flow\__init__.py                           10      0   100%
pymc4\flow\executor.py                          262     15    94%
pymc4\flow\meta_executor.py                      40      6    85%
pymc4\flow\posterior_predictive_executor.py      34      2    94%
pymc4\flow\transformed_executor.py               56      4    93%
pymc4\forward_sampling.py                        94      2    98%
pymc4\gp\__init__.py                              5      0   100%
pymc4\gp\cov.py                                  46      1    98%
pymc4\gp\gp.py                                   70      8    89%
pymc4\gp\mean.py                                 36      1    97%
pymc4\gp\util.py                                 24     19    21%
pymc4\inference\__init__.py                       2      0   100%
pymc4\inference\sampling.py                      91      6    93%
pymc4\inference\utils.py                         21      1    95%
pymc4\scopes.py                                  44      3    93%
pymc4\utils.py                                   76     13    83%
-----------------------------------------------------------------
TOTAL                                          1697    185    89%
```

</details>

File: pymc4/distributions/multivariate.py [hunk 1]
```diff
@@ -195,7 +196,11 @@ def __init__(self, name, loc, covariance_matrix, **kwargs):
     @staticmethod
     def _init_distribution(conditions):
         loc, covariance_matrix = conditions["loc"], conditions["covariance_matrix"]
-        return tfd.MultivariateNormalFullCovariance(loc=loc, covariance_matrix=covariance_matrix)
+        try:
+            chol_cov_matrix = tf.linalg.cholesky(covariance_matrix)
+        except tf.errors.InvalidArgumentError:
+            raise ValueError("Cholesky decomposition failed! Check your `covariance_matrix`.")
```
Comment (fonnesbeck) [inline pymc4/distributions/multivariate.py]: Is this guaranteed to be the case in an InvalidArgumentError?

File: pymc4/gp/cov.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+        number of rightmost dims to include in kernel computation
+    kwargs : optional
+        Other keyword arguments that tfp's ``ExponentiatedQuadratic`` kernel takes
+    """
+
+    def __init__(self, amplitude, length_scale, feature_ndims=1, **kwargs):
+        if not tf.reduce_all(amplitude > 0.0):
+            raise ValueError("`amplitude` parameter must not contains zero or negative entries.")
+        self._amplitude = amplitude
+        self._length_scale = length_scale
```
Comment (fonnesbeck) [inline pymc4/gp/cov.py]: It will be important to include ARD using the FeatureScaled kernel. Ideally this would be automated such that a scalar length scale implies ARD is not desired, while an array-valued parameter invokes ARD.

File: pymc4/gp/gp.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
+"""
+Gaussian Process Models present in PyMC4's Gaussian Process module.
```
Comment (fonnesbeck) [inline pymc4/gp/gp.py]: Somewhat redundant label

File: pymc4/gp/gp.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+from .mean import Zero
+from ..distributions import MvNormal, Normal
+from .util import stabilize
+
+
+__all__ = ["LatentGP"]
+
+
+class BaseGP:
+    def __init__(self, mean_fn=Zero(1), cov_fn=ExpQuad(1.0, 1.0, 1)):
```
Comment (fonnesbeck) [inline pymc4/gp/gp.py]: Not necessarily opposed to making a zero mean function the default, but there isnt much benefit to assume a default covariance function.

File: pymc4/gp/mean.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+    Parameters
+    ----------
+    mean1 : callable, pm.Mean
+        First mean function
+    mean2 : callable, pm.Mean
+        Second mean function
+    """
+
+    def __init__(self, mean1, mean2):
```
Comment (fonnesbeck) [inline pymc4/gp/mean.py]: No feature_ndim warning here?

File: pymc4/distributions/multivariate.py (hunk 1 above)
Comment (tirthasheshpatel) [reply pymc4/distributions/multivariate.py]: It throws ``InvalidArgumentError`` for float datatypes while ``NotNoundError`` for integer and other data types. Do you think we should be handling data types other than floats?

File: pymc4/gp/cov.py (hunk 2 above)
Comment (tirthasheshpatel) [reply pymc4/gp/cov.py]: I have added it in the base ``Covariance`` class as it is applicable to all the kernels... Should we keep the ``scale_diag`` parameter an optional keyword argument or required argument?

**Comment by tirthasheshpatel** [comment]
I wonder why the tests haven't started yet... passes on my local pc though

**Comment by lucianopaz** [comment]
What's the status with this PR? @aloctavodia, @fonnesbeck? Does this only need a rebase ame would be ready for merging?

**Comment by fonnesbeck** [comment]
Will try to look at it again tonight

**Comment by fonnesbeck** [comment]
Please go ahead and remove your local path from the notebook.

File: pymc4/gp/cov.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        return self._kernel.apply(X1, X2, **kwargs)
+
+    def __add__(self, cov2):
+        return CovarianceAdd(self, cov2)
+
+    def __mul__(self, cov2):
+        return CovarianceProd(self, cov2)
+
+    @property
+    def feature_ndims(self):
```
Comment (bwengals) [inline pymc4/gp/cov.py:69]: Much better name than `input_dim`

File: pymc4/gp/cov.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+    length_scale : tensor, array-like
+        The :math:`l` parameter of the RBF kernel
+    feature_ndims : int, optional
+        number of rightmost dims to include in kernel computation
+    kwargs : optional
+        Other keyword arguments that tfp's ``ExponentiatedQuadratic`` kernel takes
+    """
+
+    def __init__(self, amplitude, length_scale, feature_ndims=1, **kwargs):
+        self._amplitude = amplitude
```
Comment (bwengals) [inline pymc4/gp/cov.py:150]: The reason I didn't include an amplitude or scale parameter in the covariance functions in PyMC3 is because it clutters the notation for product kernels.  The user has to fix all the amplitudes at 1.0 except one of them or else it's overparameterized.  

    cov = cov1(amplitude=eta**2, ...) * cov2(amplitude=1.0, ...) * cov3(amplitude=1.0)

vs.

    cov = eta**2 * cov1(...) * cov2(...) * cov3(...)

I can see how it's cleaner than having to multiply a single cov by some scalar parameter though, so I think it would be best to either not have the amplitude parameter or use a default amplitude of 1.0.

File: pymc4/gp/gp.py [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+
+        # Let's define out GP model and its parameters
+        mean_fn = pm.gp.mean.Zero(feature_ndims=2)
+        cov_fn = pm.gp.cov.ExpQuad(1., 1., feature_ndims=2)
+        gp = pm.gp.LatentGP(mean_fn, cov_fn)
+
+        @pm.model
+        def gpmodel():
+            prior = yield gp.prior('prior', X)
+            cond = yield gp.conditional('cond', Xnew, given={'X': X, 'f': prior})
```
Comment (bwengals) [inline pymc4/gp/gp.py:75]: I'm a bit confused by conditioning on the prior?  For prediction you condition on the posterior.

**Comment by bwengals** [comment]
This looks very cool!  I added my 2 cents above, hope that's helpful.

File: pymc4/gp/cov.py (hunk 7 above)
Comment (tirthasheshpatel) [reply pymc4/gp/cov.py:150]: True! Will have to add a ``__rmul__`` and ``__radd__`` methods to do that + some tests. I will do this by today!

File: pymc4/gp/gp.py (hunk 8 above)
Comment (tirthasheshpatel) [reply pymc4/gp/gp.py:75]: Yeah! I called it ``prior`` in a sense that we are using the gp prior to infer the conditional. Will change it though as the names ``f`` and ``fcond`` sound more intuitive

File: pymc4/gp/cov.py (hunk 7 above)
Comment (tirthasheshpatel) [reply pymc4/gp/cov.py:150]: This is a little tricky unless tensorflow_probability supports white noise kernel. I have made a request to implement the white noise kernel in tensorflow_probability [tensorflow/probability#852](https://github.com/tensorflow/probability/issues/852). Without this, I will have to drop the ``_kernel`` object and implement the call method to call the individual covariance functions individually. Doesn't seem that bad but it would be nice if tfp implemented a white noise kernel.

Comment (tirthasheshpatel) [reply pymc4/gp/cov.py:173]: @bwengals I have added the support for ``__radd__`` and ``__rmul__``. I will add tests once everyone agrees with the current design, If tensorflow_probability ever adds the white noise kernel then this design can be easily refactored to take that into account. But temporarily, this is what I have ended up with...

File: pymc4/gp/gp.py (hunk 8 above)
Comment (bwengals) [reply pymc4/gp/gp.py:75]: Ah I see now, it does work the same way.  The naming threw me off.  Agree w the change

File: pymc4/gp/cov.py (hunk 7 above)
Comment (bwengals) [reply pymc4/gp/cov.py:173]: Take your time!  

Right I see, you do need an amplitude parameter without `__radd__` and `__rmul__`.  How come a white noise kernel is needed?

Comment (tirthasheshpatel) [reply pymc4/gp/cov.py:173]: Oh, sorry! I wanted to say the constant kernel. Actually, I have designed the covariance class such that the ``_init_kernel`` method initializes a kernel which is an instance of tfp's ``PositiveSemidefiniteKernel`` class. But currently, tfp doesn't support combination of covariance with non-covariance functions. So, we would need something like a constant kernel in which we can wrap the non-covariance kernel and use the ``_init_kernel`` method (and ``__call__`` method) to take care of everything else.

Comment (bwengals) [reply pymc4/gp/cov.py:173]: Oh okay I see what you're saying now, makes sense.

**Comment by twiecki** [comment]
@bwengals Thanks for reviewing. Is it OK if I assigned this to you to decide when it's ready to merge?

**Comment by tirthasheshpatel** [comment]
Greetings @bwengals,

I have added the ``__radd__`` and ``__rmul__`` methods for avoiding overparameterizing of the covariance functions. This is still problematic because TensorFlow doesn't have any ``__tensor_wrap__`` or similar utility which makes it impossible to combine a kernel with a tensor on the left of it. i.e. ``cov = tf.constant([1.]) * pm.gp.cov.ExpQuad(1.) * pm.gp.cov.ExpQuad(1.) * ...`` is not possible currently. This is not a major problem as one can still explicitly parameterize one kernel and leave others as is. i.e. ``cov = pm.gp.cov.ExpQuad(length_scale=1., amplitude=1.) * pm.gp.cov.ExpQuad(length_scale=1.) * ...``. 

I have designed a ``__array_wrap__`` method that can be used to combine kernels with numpy arrays on the left of them. ``cov = np.array([1.]) * pm.gp.cov.ExpQuad(1.) * pm.gp.cov.ExpQuad(1.) * ...``.
 
I have also completed writing tests for all the changes and am hoping to summarize all the changes very soon so it is possible to review this huge pr.

**Comment by bwengals** [comment]
Sure @twiecki I'd be happy too, should have a chance this weekend

**Comment by tirthasheshpatel** [comment]
I don't want to change the major flow of this pr as it would probably make the job of reviewers much more difficult.
But here are a few updates from my side (hope they are helpful):
 - I want to reintroduce the ``diag`` parameter (from the discussion I had on discourse). I also want to implement the ``active_dims`` feature in a latter pr from the discussion @fonnesbeck had on tfp [tensorflow/probability#862](https://github.com/tensorflow/probability/issues/862)
 - I will try to create a separate ``Parameter`` (or something like a ``HyperParameter``) class to represent a GP hyperparameter and other utilities to start drafting some of the optimization methods.
 - The test suite is too congested as of now. I believe that PyMC3's style of writing tests provides more user-side experience than using fixtures. OTOH, fixtures help reduce a lot of redundant code. It depends on what the team wants right now, i.e: a friendly UI or functionality is the main focus. Does this need further discussion?
 - TFP doesn't support combination or kernels with non-kernels which makes it a little difficult to embed the TFP's addition interface in the proposed interface efficiently.

@fonnesbeck Do you think that it would be a nice choice to resort to TFP for almost all tasks or should we prefer to write our own code for some specific tasks?

Thank you so much to all the reviewers so far for going through this pr!!!

**Comment by fonnesbeck** [comment]
I would lean on TFP wherever it makes sense to do so. The goal here is to provide a higher-level interface to GPs than TFP provides, something more along the lines of what Bill gave us in PyMC3. I think you've done a good job of doing that so far.

File: pymc4/gp/cov.py (hunk 7 above)
Comment (fonnesbeck) [reply pymc4/gp/cov.py:203]: It may make sense to implement the ConstantKernel and use that for amplitude. It may make it clearer to users how the amplitude works, and would avoid trying to specify multiple amplitudes for multiplicative kernels. What do you think?

File: pymc4/gp/util.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+ArrayLike = Union[np.ndarray, tf.python.framework.ops.EagerTensor]
+TfTensor = tf.python.framework.ops.EagerTensor
+
+
+def stabilize(K, shift=1e-6):
+    r"""Add a diagonal shift to a covarience matrix"""
+    return tf.linalg.set_diag(K, tf.linalg.diag_part(K) + shift)
+
+
+def plot_gp_dist(
```
Comment (fonnesbeck) [inline pymc4/gp/util.py:16]: We might want plotting functions in their own submodule, rather than pooled with other utilities. There are potentially quite a few plots that could eventually be added.

**Comment by fonnesbeck** [comment]
Looks good. I don't necessarily get good estimates when I run the notebook, but that can be addressed in another PR. Good start on this!
